proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,429 @@
+   not include the ``import`` keyword. (Modules named ``_`` without an
+   non-empty export list but without the ``import`` keyword are still useful
+   as a way of declaring instances that use local definitions.)
+
+7. Every ``class``, ``data``, ``newtype``, ``data instance``, and ``newtype
+   instance`` declaration implicitly creates a new local module. The name of
+   the local module matches the name of the declared type. All entities (e.g.,
+   method names, constructors, record selectors) brought into scope within the
+   declaration, including the type itself, are put into this local module.
+   
+   If the pseudo-keyword ``qualified`` appears directly after the keyword(s)
+   that begin the declaration, these internal definitions are not brought into
+   the outer scope. Otherwise, they are (just like usual). Exception: the type
+   itself is always brought into scope unqualified. This feature is enabled
+   only when ``-XLocalModules`` is in effect.
+
+   Associated ``data`` and ``newtype`` instances create modules at the level
+   of the enclosing ``instance`` declaration: the ``data``\/\ ``newtype``
+   module is *not* nested within the class module.
+
+8. Local modules may be extended via the declaration of another local module
+   of the same name. If local modules with the same name are in scope at the
+   same time (either through importation or declaration) their contents are
+   simply merged. Individual identifiers that are multiply defined will be
+   an error if used ambiguously.
+
+9. A new declaration form is introduced with the following BNF::
+
+     decl ::= ... | 'import' 'module' conid [ import_spec ]
+
+   This declaration takes all entities (or those entities named in the
+   ``import_spec``) in scope with a *conid*\ ``.`` and brings them into scope
+   into the local context unqualified.
+
+   The declaration is allowed in ``let`` and ``where`` clauses.
+
+   Note that the declaration form includes the word ``module`` to distinguish
+   it from a normal ``import`` which induces a dependency on another file. An
+   ``import module`` declaration cannot induce a dependency.
+
+   This declaration form is allowed only with ``-XLocalModules``.
+
+10. Local modules may be imported with this import item::
+
+      import_item ::= ... | [ 'import' ] 'module' conid [ import_spec ]
+
+    This brings the local module named *conid* into scope. It is an error if
+    this import item appears in a list of imports from a module that does not
+    export the local module *conid*.
+
+    Also brought into scope are the entities listed in the *import_spec*;
+    these are brought into scope qualified with the *conid*\ ``.`` prefix. If
+    the *import_spec* is not provided, then all entities in *conid* are
+    brought into scope (qualified).
+
+    If the ``import`` keyword is included, then all entities brought into
+    scope qualified are also brought into scope unqualified.
```

Comment:
Does this unqualify one level of module nesting depth? If `Foo.hs` exports module `M1` which exports `M2` which exports `bar`, and I write
```
import Foo ( import module M1 )
```
is `M2.bar` or `bar` in scope?